### ISSUE TITLE
2.6.10 dev SET name and GET name oddness

### ISSUE THREAD

**Issue by OverlanAI** [Program-O/Program-O#408]
Seems this sometimes works and other times fails.
I will set a value like set name="she" SOMEVALUE and sometimes when i get name "she" it returns the name i set correctly but sometimes it will return unknown or in a few tests it will just say the word "she" and not the value using the same test i made.
I was not expecting this to happen and ran into it checking for a non related problem.
Is there some AIML test someone has so i can test for this problem better?

TNX

**Comment by Dave-Morton** [comment]
@tringate has written a comprehensive AIML test suite for not just this, but for a wide range of other situations and scenarios, as well. I'm sure that now that I've tagged him in this issue he should stop by and let you know the best way to obtain the AIML set.

Personally, I think it would be much easier to see the AIML code that you're using so that we can rule it out as a potential cause for the symptoms you're seeing.

**Comment by OverlanAI** [comment]
i think i may see the problem.
Seems it does not like set name="she" somevalue if its inside of RANDOM tags.
In my case i have list of names it picks from randomly and i have it also set name="she" to whatever it picked. It does show the name it picks correctly in chat but it never sets the value even though i have A set name="her"NAME for each name in the random list.
Basically when i call my srai it should pick and set one of the names randomly and return.
So maybe its just the way i am trying to do it?
If is set just one name then get the name then it is correct but never if it is set randomly it seems.
Anyway ill attach something if it should work that way but it is not.
tnx

**Comment by OverlanAI** [comment]
![set npc name random](https://user-images.githubusercontent.com/36704725/39279524-b53400a8-48c7-11e8-81ef-913d609b271c.PNG)

Really small so this seems simpler to show.
Basically the convo would be
client: Who told you that?
AI: Monica told me.
but i also expected it to set the name Monica as name="she".
I think you get the idea from this.
Could be that i am just doing that wrong or maybe it is ok and something else is not working as it should.

**Comment by tringate** [comment]
I have tested your AIML and it works perfectly for me on my system.

Here is the output of several tests using your exact code:
Slow Putzo: RANDOM TEST
April Genie: Angel
Slow Putzo: RANDOM TEST
April Genie: Lexi
Slow Putzo: RANDOM TEST
April Genie: Barbarella
Slow Putzo: RANDOM TEST
April Genie: Tiggy
Slow Putzo: RANDOM TEST
April Genie: Tiggy
Slow Putzo: RANDOM TEST
April Genie: Tiggy
Slow Putzo: RANDOM TEST
April Genie: Barbarella

Here are the two AIML categories I used to test your routine:
```
<category><pattern>RANDOM TEST</pattern><template>
<srai>SET NPC NAME</srai>
<get name="she"/>
</template></category>



<category><pattern>SET NPC NAME</pattern><template>
<think>
<random>
<li><set name="she">Monica</set></li>
<li><set name="she">Barbarella</set></li>
<li><set name="she">Tiggy</set></li>
<li><set name="she">Lexi</set></li>
<li><set name="she">Angel</set></li>
</random>
</think>
</template></category>`

**Comment by tringate** [comment]
Your question pointed out to me that I had not written any test cases for the RANDOM tag.  There are so many other tags that were failing that I never got to those which were working without issues.

Here is a zip file of the current AIML test cases.  four of them are still failing and produce easy to read failure information on which case didn't run correctly.  You execute the suite of test cases by using the input of "QUICKTEST"
[_Test-cases-2018-04-05a.zip](https://github.com/Program-O/Program-O/files/1949724/_Test-cases-2018-04-05a.zip)

**Comment by OverlanAI** [comment]
Ahh, thank you.
There is one difference between what i have and what you have is yours is wrapped in "think tags".
I thought of doing that but from my understanding, think tags would just suppress the display of the name it would have picked so i did not see the need.
I will wrap mine in the same way and also run your tests because it also exhibited problems randomly.
Anyway at least i had it partly right lol.
tnx

**Comment by OverlanAI** [comment]
well it is still returning "undefined" using exactly what you have above.
Really odd.
If I just flat out set name="she" then it will display the value of what I set with a get name="she".
Pass the aspirin please.

**Comment by OverlanAI** [comment]
Okay, i did find that if it use the admin json gui and say "random test" it does work ok.
But when i use HTTP to say the same thing if fails and get the odd results.
I changed my HTTP method from GET to POST and i now get the correct result so far via HTTP POST.
So seems like some of what i am running into is not AIML related at all.

Could it be timing?

**Comment by OverlanAI** [comment]
Well it still acts odd but not as bad. I have changed the name="she" to name="npc" to hold the random pick in. 
It would go along ok but after awhile get name="she" started returning "undefined" again.
name="she" could be getting cleared somehow so i am continuing the test using its own var name as NPC not SHE to be sure.
tnx (dont close this just yet please)

**Comment by OverlanAI** [comment]
In 2.6.10 dev with HTTP post it does work but not well (still get the oddness over time)
In 2.6.9 This does not work at all in random tag set but does if it is a direct set and could care less about HTTP method.
On looking thru working ALICE 1.9 as ref. for some other example use of what i was trying to do, I found none that try to set from with in random tags.
So at best i can say it kind of works sometimes but seems not a valid way to go about it to start with.

TNX (the operation was a success but the patent died)

**Comment by OverlanAI** [comment]
TC-08 is a get bot domain value from the personality db. 
As for the other fails it is hard to say still. It is running as part of the ALICE 1.9 (master set) so she might be playing tricks on me running the tests.
On the other hand in ALICE i see most of your test cases used in actual functions as part of there set.
What all that means to you IDK but that is the result i got here.
In your email you seem to feel that the test's will not play well with others but i find that a bit odd that it should not matter. I can reset to OM but that woun't do anything with static data.
Anyway thats the result of it at least trying to play along will with ALICE but she gets in a bad mood sometimes lol.
tnx

**Comment by OverlanAI** [comment]
I will add in a POP OM (directly clears stack) and also clear profile data in the same way in your AIML QUICKEST and see what i get.
I have been fighting stack issues for along time now but it is getting more clear to me what is going on.
Ill re-post a new test soon.

**Comment by OverlanAI** [comment]
LETS TALK ABOUT *
`<set name="it"><set name="topic"><person/></set></set> is an interesting topic.`
This does not work at all, person becomes null so the topic never gets set.
But if i just do this.
`<set name="topic"><person/></set> is an interesting topic.`
then the topic is set correctly.

The bad news is the nested set's are real common in ALICE 1.9 (master) and other bots and with out that working you end up with a real broken bot.
In pre-patched 2.6.10 it worked ok with the nested sets.
I will have to backtrack to see what patch broke this and other problems like this i am finding in current dev.
Might want to add that to your test stuff or at least check it on your system and see what you get.
TNX

**Comment by OverlanAI** [comment]
yep all good stuff and real tests but i guess that ass dont care so see you on the fork.
